system/lib/libc/dynlink.c

```diff
@@ -105,7 +108,236 @@ static struct dso* load_library_start(const char* name, int flags) {
+  // with sync_one_onerror.
+  pthread_rwlock_rdlock(&lock);
+  if (!thread_local_tail) {
+    thread_local_tail = head;
+  }
+  sync_next(thread_local_tail->next, promise_id);
+}
+
+// Called on background threads to synchronously "catch up" with all the DSOs
+// that are currently loaded.
+bool _emscripten_thread_sync_code() {
+  // Should only ever be called from a background thread.
+  assert(!emscripten_is_main_runtime_thread());
+  ensure_init();
+  if (thread_local_tail == tail) {
+#ifdef DYLINK_DEBUG
+    _emscripten_errf("%p: emscripten_thread_sync_code: already in sync", pthread_self());
+#endif
+    return true;
+  }
+  pthread_rwlock_rdlock(&lock);
+  if (!thread_local_tail) {
+    thread_local_tail = head;
+  }
+  while (thread_local_tail->next) {
+    struct dso* p = thread_local_tail->next;
+#ifdef DYLINK_DEBUG
+    _emscripten_errf(
+      "%p: emscripten_thread_sync_code: %s mem_addr=%p mem_size=%zu "
+      "table_addr=%p table_size=%zu",
+      pthread_self(),
+      p->name,
+      p->mem_addr,
+      p->mem_size,
+      p->table_addr,
+      p->table_size);
+#endif
+    void* success = _dlopen_js(p);
+    if (!success) {
+      pthread_rwlock_unlock(&lock);
+      _emscripten_errf("%p: emscripten_thread_sync_code failed: %s", pthread_self(), dlerror());
+      return false;
+    }
+    thread_local_tail = p;
+  }
+  pthread_rwlock_unlock(&lock);
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: emscripten_thread_sync_code done", pthread_self());
+#endif
+  return true;
+}
+
+// This structure exists so that we can stash the status of work we do
+// during emscripten_proxy_async_with_callback.
+// If the function that emscripten_proxy_async_with_callback ran included
+// a return value we could avoid this struct and the malloc/free of it.
+// See https://github.com/emscripten-core/emscripten/issues/18378
+struct thread_sync_data {
+  int promise_id;
+  bool result;
+};
+
+static void do_thread_sync(void* arg) {
+  struct thread_sync_data* data = (struct thread_sync_data*)arg;
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: do_thread_sync: %d", pthread_self(), data->promise_id);
+#endif
+  data->result = _emscripten_thread_sync_code();
+}
+
+// Called once _emscripten_proxy_sync_code completes
+static void done_thread_sync(void* arg) {
+  struct thread_sync_data* data = (struct thread_sync_data*)arg;
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: done_thread_sync: promise_id=%d success=%d",
+                   pthread_self(),
+                   data->promise_id,
+                   data->result);
+#endif
+  if (data->result) {
+    _emscripten_promise_resolve(data->promise_id, NULL);
+  } else {
+#if ABORT_ON_SYNC_FAILURE
+    abort();
+#else
+    _emscripten_promise_reject(data->promise_id);
+#endif
+  }
+  free(data);
+}
+
+static em_proxying_queue* dlopen_proxying_queue;
+
+__attribute__((constructor)) static void create_queue() {
+  dlopen_proxying_queue = em_proxying_queue_create();
+}
```

Comment:
Can this be done lazily the first time anyone calls dlopen? That seems cleaner than doing it in a static constructor.

Reply:
Done.  Since this is global shared between many threads I did end up having to make the pointer atomic.